Setup RabbitMQ

doomspork opened the pull request.

Commit:
Setup RabbitMQ

Commit:
Clean up gRPC server and add test coverage

Commit:
Support build updates and improve pid management

Commit:
Update Bottle SHA

Commit:
Expand on build attributes

Add model and order id to all builds. Additionally, support the INPROGRESS
status.

Commit:
Ignore /lib

Commit:
Update Cachex usage

Commit:
Add logging

Commit:
Emit message on build updates

Commit:
Add handler logging

Commit:
Improvements

Commit:
Use hal_id for logs

Commit:
Ignore messages from self

Commit:
Update builds to work off HAL ids

Commit:
Add missing component support

Since this is easy to calculate we won't persist it by rather keep it in
state. Following this we can update the gRPC call to return the missing
components for display in the UI.

btkostner commented on lib/assembly/schemas/build.ex:
```diff
@@ -14,6 +14,8 @@ defmodule Assembly.Schemas.Build do
     field :order_id, :string
     field :status, BuildStatusEnum
 
+    field :missing_components, :any, virtual: true
```
Can we switch this to be a bit more typed? It makes it a bit easier to read when looking at the schema.
```
field :missing_components, {:array, :string}, virtual: true
```

doomspork replied:
It's not a string, so it would need to be `:map` but it's actually a struct, but not an association since it's virtual. Therefore `:any` is the only way to skip type checking which we're not looking for.

btkostner replied:
hmmm, that's annoying. I'm assuming you can't do `{:map, %Bottle...}` so I guess this is the best we can do.

doomspork commented on lib/assembly/server.ex:
```diff
@@ -1,24 +1,33 @@
-    BuildListResponse.new(builds: builds)
+  @spec build_list(BuildListRequest.t(), GRPC.Server.Stream.t()) :: any
+  def build_list(_request, stream) do
+    query =
+      from b in Schemas.Build,
+        left_join: c in assoc(b, :build_components),
+        where: b.status != :built,
+        preload: [build_components: c]
```
Right now this is based on DB lookup. I believe we want to shift to the supervisor here instead. It's less hits to the DB and it's more updated info than the DB as the persistence may not have occurred yet.

Commit:
Update gRPC payload and data source

Add support for missing components, update our inquiries into builds
from the DB to the supervisor

Commit:
Logging updates

Commit:
Ensure missing components is populated

Commit:
Remove failing test

This test wasn't great to begin with and now it's resulting in a suite
failure. We should devise a better test for this.
